File: src/lib/Backend/derived/siteAnalytics.ts
import type { AesoHourlyMarketRecord } from "../aeso/aesoRepository";
import type { ScadaHourlyRecord } from "../scada/scadaRepository";

export type CurtailmentSiteRequest = {
  latitude: number;
  longitude: number;
  startIso: string;
  endIso: string;
  turbineCount: number;
};

export type CurtailmentRow = {
  timestamp: string;
  poolPriceCadPerMWh: number;
  actualGenerationMWh: number;
  modeledGenerationMWh: number;
  curtailmentGapMWh: number;
};

export type LoadStorageRequest = CurtailmentSiteRequest & {
  batteryPowerMw: number;
  batteryDurationHours: number;
  roundTripEfficiency: number;
};

export type NetworkFiberRequest = {
  siteId: string;
  inferenceMaxLatencyMs: number;
  trainingMaxLatencyMs: number;
};

export function buildCurtailmentView(
  request: CurtailmentSiteRequest,
  marketRows: AesoHourlyMarketRecord[],
  scadaRows: ScadaHourlyRecord[],
): { rows: CurtailmentRow[]; summary: { hours: number; totalGapMWh: number } } {
  validateCurtailmentRequest(request);

  const rows = buildCurtailmentRows(marketRows, scadaRows, request.turbineCount);
  const totalGapMWh = rows.reduce((sum, row) => sum + row.curtailmentGapMWh, 0);

  return {
    rows,
    summary: {
      hours: rows.length,
      totalGapMWh,
    },
  };
}

export function simulateLoadStorage(
  request: LoadStorageRequest,
  marketRows: AesoHourlyMarketRecord[],
  scadaRows: ScadaHourlyRecord[],
): { capturedMWh: number; releasedMWh: number; estimatedGrossRevenueCad: number } {
  validateCurtailmentRequest(request);
  assertFinitePositive(request.batteryPowerMw, "Invalid batteryPowerMw");
  assertFinitePositive(request.batteryDurationHours, "Invalid batteryDurationHours");
  if (request.roundTripEfficiency <= 0 || request.roundTripEfficiency > 1) {
    throw new Error("Invalid roundTripEfficiency");
  }

  const { rows } = buildCurtailmentView(request, marketRows, scadaRows);
  const storageMWh = request.batteryPowerMw * request.batteryDurationHours;

  let stateOfChargeMWh = 0;
  let capturedMWh = 0;
  let releasedMWh = 0;
  let estimatedGrossRevenueCad = 0;

  for (const row of rows) {
    const availableHeadroomMWh = Math.max(0, storageMWh - stateOfChargeMWh);
    const chargeMWh = Math.min(row.curtailmentGapMWh, request.batteryPowerMw, availableHeadroomMWh);
    stateOfChargeMWh += chargeMWh;
    capturedMWh += chargeMWh;

    if (row.poolPriceCadPerMWh > 0 && stateOfChargeMWh > 0) {
      const dischargeMWh = Math.min(stateOfChargeMWh, request.batteryPowerMw);
      stateOfChargeMWh -= dischargeMWh;

      const deliveredMWh = dischargeMWh * request.roundTripEfficiency;
      releasedMWh += deliveredMWh;
      estimatedGrossRevenueCad += deliveredMWh * row.poolPriceCadPerMWh;
    }
  }

  return {
    capturedMWh,
    releasedMWh,
    estimatedGrossRevenueCad,
  };
}

export function evaluateNetworkFiber(
  request: NetworkFiberRequest,
  distanceToPopKm: number,
): {
  distanceToPopKm: number;
  estimatedLatencyMs: number;
  workloads: { inference: boolean; training: boolean };
} {
  if (!request.siteId.trim()) {
    throw new Error("Invalid siteId");
  }
  assertFinitePositive(request.inferenceMaxLatencyMs, "Invalid inferenceMaxLatencyMs");
  assertFinitePositive(request.trainingMaxLatencyMs, "Invalid trainingMaxLatencyMs");
  assertFinitePositive(distanceToPopKm, "Invalid distanceToPopKm");

  const estimatedLatencyMs = Number((distanceToPopKm * 0.005).toFixed(3));

  return {
    distanceToPopKm,
    estimatedLatencyMs,
    workloads: {
      inference: estimatedLatencyMs <= request.inferenceMaxLatencyMs,
      training: estimatedLatencyMs <= request.trainingMaxLatencyMs,
    },
  };
}

function buildCurtailmentRows(
  marketRows: AesoHourlyMarketRecord[],
  scadaRows: ScadaHourlyRecord[],
  turbineCount: number,
): CurtailmentRow[] {
  assertFinitePositive(turbineCount, "Invalid turbineCount");

  const scadaByTimestamp = new Map(scadaRows.map((row) => [row.timestamp, row]));
  const rows: CurtailmentRow[] = [];

  for (const marketRow of marketRows) {
    const scadaRow = scadaByTimestamp.get(marketRow.timestamp);
    if (!scadaRow) {
      continue;
    }

    const modeledPerTurbineMWh = modeledOutputPerTurbine(scadaRow.windSpeedMs);
    const modeledGenerationMWh = modeledPerTurbineMWh * turbineCount;
    const curtailmentGapMWh = Math.max(0, modeledGenerationMWh - marketRow.actualGenerationMWh);

    rows.push({
      timestamp: marketRow.timestamp,
      poolPriceCadPerMWh: marketRow.poolPriceCadPerMWh,
      actualGenerationMWh: marketRow.actualGenerationMWh,
      modeledGenerationMWh,
      curtailmentGapMWh,
    });
  }

  return rows.sort((a, b) => a.timestamp.localeCompare(b.timestamp));
}

function modeledOutputPerTurbine(windSpeedMs: number): number {
  if (!Number.isFinite(windSpeedMs) || windSpeedMs <= 5) {
    return 0;
  }
  return Math.min(windSpeedMs - 5, 5);
}

function validateCurtailmentRequest(request: CurtailmentSiteRequest) {
  if (!Number.isFinite(request.latitude) || request.latitude < -90 || request.latitude > 90) {
    throw new Error("Invalid latitude");
  }
  if (!Number.isFinite(request.longitude) || request.longitude < -180 || request.longitude > 180) {
    throw new Error("Invalid longitude");
  }

  const start = Date.parse(request.startIso);
  const end = Date.parse(request.endIso);
  if (!Number.isFinite(start) || !Number.isFinite(end) || start > end) {
    throw new Error("Invalid date range");
  }

  const maxRangeMs = 366 * 24 * 60 * 60 * 1000;
  if (end - start > maxRangeMs) {
    throw new Error("Date range too large");
  }

  assertFinitePositive(request.turbineCount, "Invalid turbineCount");
}

function assertFinitePositive(value: number, message: string) {
  if (!Number.isFinite(value) || value <= 0) {
    throw new Error(message);
  }
}

File: src/backend/src/analytics/site_analytics.py
"""Site analytics - curtailment, load storage, and network fiber analysis."""
from typing import Any
from datetime import datetime, timedelta
from ..models.requests import CurtailmentSiteRequest, LoadStorageRequest, NetworkFiberRequest
from ..models.responses import AesoHourlyMarketRecord, ScadaHourlyRecord


class CurtailmentRow:
    """Single row in curtailment analysis."""
    def __init__(
        self,
        timestamp: str,
        pool_price_cad_per_mwh: float,
        actual_generation_mwh: float,
        modeled_generation_mwh: float,
        curtailment_gap_mwh: float
    ):
        self.timestamp = timestamp
        self.poolPriceCadPerMWh = pool_price_cad_per_mwh
        self.actualGenerationMWh = actual_generation_mwh
        self.modeledGenerationMWh = modeled_generation_mwh
        self.curtailmentGapMWh = curtailment_gap_mwh
    
    def to_dict(self) -> dict[str, Any]:
        return {
            "timestamp": self.timestamp,
            "poolPriceCadPerMWh": self.poolPriceCadPerMWh,
            "actualGenerationMWh": self.actualGenerationMWh,
            "modeledGenerationMWh": self.modeledGenerationMWh,
            "curtailmentGapMWh": self.curtailmentGapMWh,
        }


def build_curtailment_view(
    request: CurtailmentSiteRequest,
    market_rows: list[AesoHourlyMarketRecord],
    scada_rows: list[ScadaHourlyRecord],
) -> dict[str, Any]:
    """Build curtailment analysis view with gap calculations."""
    validate_curtailment_request(request)
    
    rows = build_curtailment_rows(market_rows, scada_rows, request.turbineCount)
    total_gap_mwh = sum(row.curtailmentGapMWh for row in rows)
    
    return {
        "rows": [row.to_dict() for row in rows],
        "summary": {
            "hours": len(rows),
            "totalGapMWh": total_gap_mwh,
        }
    }


def simulate_load_storage(
    request: LoadStorageRequest,
    market_rows: list[AesoHourlyMarketRecord],
    scada_rows: list[ScadaHourlyRecord],
) -> dict[str, float]:
    """Simulate battery storage capture and release."""
    validate_curtailment_request(request)
    assert_finite_positive(request.batteryPowerMw, "Invalid batteryPowerMw")
    assert_finite_positive(request.batteryDurationHours, "Invalid batteryDurationHours")
    
    if request.roundTripEfficiency <= 0 or request.roundTripEfficiency > 1:
        raise ValueError("Invalid roundTripEfficiency")
    
    curtailment = build_curtailment_view(request, market_rows, scada_rows)
    rows = [CurtailmentRow(**r) for r in curtailment["rows"]]
    
    storage_mwh = request.batteryPowerMw * request.batteryDurationHours
    state_of_charge_mwh = 0.0
    captured_mwh = 0.0
    released_mwh = 0.0
    estimated_gross_revenue_cad = 0.0
    
    for row in rows:
        # Charge battery from curtailment
        available_headroom_mwh = max(0, storage_mwh - state_of_charge_mwh)
        charge_mwh = min(row.curtailmentGapMWh, request.batteryPowerMw, available_headroom_mwh)
        state_of_charge_mwh += charge_mwh
        captured_mwh += charge_mwh
        
        # Discharge battery when price is positive
        if row.poolPriceCadPerMWh > 0 and state_of_charge_mwh > 0:
            discharge_mwh = min(state_of_charge_mwh, request.batteryPowerMw)
            state_of_charge_mwh -= discharge_mwh
            
            delivered_mwh = discharge_mwh * request.roundTripEfficiency
            released_mwh += delivered_mwh
            estimated_gross_revenue_cad += delivered_mwh * row.poolPriceCadPerMWh
    
    return {
        "capturedMWh": captured_mwh,
        "releasedMWh": released_mwh,
        "estimatedGrossRevenueCad": estimated_gross_revenue_cad,
    }


def evaluate_network_fiber(
    request: NetworkFiberRequest,
    distance_to_pop_km: float,
) -> dict[str, Any]:
    """Evaluate network fiber latency and workload suitability."""
    if not request.siteId.strip():
        raise ValueError("Invalid siteId")
    
    assert_finite_positive(request.inferenceMaxLatencyMs, "Invalid inferenceMaxLatencyMs")
    assert_finite_positive(request.trainingMaxLatencyMs, "Invalid trainingMaxLatencyMs")
    assert_finite_positive(distance_to_pop_km, "Invalid distanceToPopKm")
    
    estimated_latency_ms = round(distance_to_pop_km * 0.005, 3)
    
    return {
        "distanceToPopKm": distance_to_pop_km,
        "estimatedLatencyMs": estimated_latency_ms,
        "workloads": {
            "inference": estimated_latency_ms <= request.inferenceMaxLatencyMs,
            "training": estimated_latency_ms <= request.trainingMaxLatencyMs,
        }
    }


def build_curtailment_rows(
    market_rows: list[AesoHourlyMarketRecord],
    scada_rows: list[ScadaHourlyRecord],
    turbine_count: int,
) -> list[CurtailmentRow]:
    """Build curtailment rows by joining market and SCADA data."""
    assert_finite_positive(turbine_count, "Invalid turbineCount")
    
    # Create lookup map for SCADA data
    scada_by_timestamp = {
        row.timestamp.isoformat() + "Z": row
        for row in scada_rows
    }
    
    rows = []
    for market_row in market_rows:
        timestamp_key = market_row.timestamp.isoformat() + "Z"
        scada_row = scada_by_timestamp.get(timestamp_key)

        if not scada_row:
            continue

        modeled_per_turbine_mwh = modeled_output_per_turbine(scada_row.windSpeedMs)
        modeled_generation_mwh = modeled_per_turbine_mwh * turbine_count
        curtailment_gap_mwh = max(0, modeled_generation_mwh - market_row.actualGenerationMWh)

        rows.append(CurtailmentRow(
            timestamp=timestamp_key,
            pool_price_cad_per_mwh=market_row.poolPriceCadPerMWh,
            actual_generation_mwh=market_row.actualGenerationMWh,
            modeled_generation_mwh=modeled_generation_mwh,
            curtailment_gap_mwh=curtailment_gap_mwh,
        ))

    # Sort by timestamp
    rows.sort(key=lambda r: r.timestamp)
    return rows


def modeled_output_per_turbine(wind_speed_ms: float) -> float:
    """Simple wind speed to power model."""
    if not isinstance(wind_speed_ms, (int, float)) or wind_speed_ms <= 5:
        return 0.0
    return min(wind_speed_ms - 5, 5)


def validate_curtailment_request(request: CurtailmentSiteRequest) -> None:
    """Validate curtailment request parameters."""
    if not isinstance(request.latitude, (int, float)) or request.latitude < -90 or request.latitude > 90:
        raise ValueError("Invalid latitude")

    if not isinstance(request.longitude, (int, float)) or request.longitude < -180 or request.longitude > 180:
        raise ValueError("Invalid longitude")

    # Parse ISO timestamps
    try:
        start = datetime.fromisoformat(request.startIso.replace('Z', '+00:00'))
        end = datetime.fromisoformat(request.endIso.replace('Z', '+00:00'))
    except (ValueError, AttributeError):
        raise ValueError("Invalid date range")

    if start > end:
        raise ValueError("Invalid date range")

    max_range = timedelta(days=366)
    if end - start > max_range:
        raise ValueError("Date range too large")

    assert_finite_positive(request.turbineCount, "Invalid turbineCount")


def assert_finite_positive(value: float | int, message: str) -> None:
    """Assert value is finite and positive."""
    if not isinstance(value, (int, float)) or value <= 0:
        raise ValueError(message)
